### PR TITLE
fixing sorting in stride indices

### PR DESCRIPTION
**Pull request by jjsjann123**
Updating `computeStrideProps` logic to break ties on stride_indices.

For two dimension with identical stride, the dimension with size-1 should be considered as the faster dimension. Otherwise, its stride should be the product of existing stride and the size of the other dimension.

Note that there's still inconsistency between eager memory_format and stride_properties in JIT, this is a design issue due to the ambiguity on size-1 stride. One example showing this failing test has been disabled in the added cpp test

### CODE REVIEW

**Comment by jjsjann123** [comment]
cc'ing @Krovatkin

File: aten/src/ATen/core/type.cpp [hunk 1]
```diff
@@ -1410,14 +1410,35 @@ VaryingShape<Stride> TensorType::computeStrideProps(
...
-        // i.e. (1, 1, 5)
+      [&strides, &sizes](const int& a, const int& b) {
         if (strides[a] == strides[b]) {
-          return a > b;
+          // The index order is ambiguous with 2 dimensions of size 1.
+          // In this case of uncertainty, default to descending index order.
+          if (sizes[a] == sizes[b]) {
+            return a > b;
+          } else {
+            return sizes[a] == 1;
```
Comment (dzhulgakov) [inline aten/src/ATen/core/type.cpp:1440]: should it be `sizes[a] < sizes[b]`? there technically can be a weird self-overlapping tensor with equal strides and sizes for two dimensions which are not size 1. Just to make sure that the comparison function is a proper comparison

Comment (jjsjann123) [reply aten/src/ATen/core/type.cpp:1440]: Good catch!
I'm somewhat overlooking self-overlapping cases, as it's unclear to me how to make sense of stride index for them. With self-overlapping case, I think we should just fallback to the canonical assumption `return a > b;`

The logic I have doesn't look right in the general case. Let me update it.

File: aten/src/ATen/core/type.cpp [hunk 2]
```diff
@@ -1410,14 +1410,48 @@ VaryingShape<Stride> TensorType::computeStrideProps(
...
+          // ii. when it's self-overlapping, we just default to PyTorch
+          //   canonical memory format and checks for whichever axis that comes
+          //   first;
+          if (sizes[a] == 1 && sizes[b] != 1) {
+            return true;
+          } else if (sizes[a] != 1 && sizes[b] == 1) {
+            return false;
+          } else {
+            return a > b;
+          }
```
Comment (jjsjann123) [inline aten/src/ATen/core/type.cpp:1454]: @dzhulgakov Does this look better? ^^^

**Review by dzhulgakov — APPROVED** [review]

File: aten/src/ATen/core/type.cpp [hunk 3]
```diff
@@ -1410,14 +1410,48 @@ VaryingShape<Stride> TensorType::computeStrideProps(
...
-        // i.e. (1, 1, 5)
+      [&strides, &sizes](const int& a, const int& b) {
         if (strides[a] == strides[b]) {
-          return a > b;
+          // when strides of two axes are identical:
+          // i. if it's not self-overlapping:
+          //   1. if axis_a is the only size-1 axis, we think axis_a comes
+          //      before axis_b. Since strides[b] == strides[a] * sizes[a];
+          //      Meanwhile, it is not true for axis a;
+          //   2. if axis_b is the only size-1 axis, likewise, we think axis_a
```
Comment (Krovatkin) [inline aten/src/ATen/core/type.cpp:1439]: nit: if we are using `likewise` could we say "we think axis_b comes before axis_a?

File: aten/src/ATen/core/type.cpp [hunk 4]
```diff
@@ -1410,14 +1410,48 @@ VaryingShape<Stride> TensorType::computeStrideProps(
...
+          //      comes after axis_b;
+          //   3. when both axes are size-1, we just assume canonical memory
+          //      format in PyTorch and takes whichever axis that comes first to
+          //      be the outer dimension;
+          //
+          // ii. when it's self-overlapping, we just default to PyTorch
```
Comment (Krovatkin) [inline aten/src/ATen/core/type.cpp:1445]: self-overlapping means we have more than one size-1 in this context?

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: aten/src/ATen/core/type.cpp (hunk 4 above)
Comment (jjsjann123) [reply aten/src/ATen/core/type.cpp:1445]: I was referring to the opposite.

self-overlapping == memory overlapping. In this context, when we have `strides[a] == strides[b]`, neither dimension should be size-1.

File: aten/src/ATen/core/type.cpp (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/core/type.cpp:1454]: The goal of jit should be to have consistent behavior with eager, which means that conditional here should be
```
if (sizes[a] != sizes[b]) {
  return sizes[a] < sizes[b]; //that also orders self-overlapping tensors, preferring "smaller" rolling windows
} else {
  return a > b;
}
```
https://github.com/pytorch/pytorch/blob/971c57f1d094806eab5340263dd59dd817267994/aten/src/ATen/TensorIterator.cpp#L192-L200
The orderding of `0` (broadcasted) strides to the beginning is highly suspect, and as you can see, eager avoids doing that.

File: aten/src/ATen/core/type.cpp [hunk 5]
```diff
@@ -1410,14 +1410,48 @@ VaryingShape<Stride> TensorType::computeStrideProps(
...
+          if (sizes[a] == 1 && sizes[b] != 1) {
+            return true;
+          } else if (sizes[a] != 1 && sizes[b] == 1) {
+            return false;
+          } else {
+            return a > b;
+          }
         }
         return strides[a] < strides[b];
       });
```
Comment (ngimel) [inline aten/src/ATen/core/type.cpp:1457]: Look below at line 1469 https://github.com/pytorch/pytorch/pull/63940/files#diff-adcb9abb3299ad7384628ad8133a0ffa7f09b0491cdc2346af9c024cd492eab6R1469, looks like a long standing typo (should be `contiguous_ = sizes[stride_indices[i]] == 1`), amazing that it hasn't been caught.

**Comment by jjsjann123** [comment]
@ngimel Updated the computation per TensorIterator logic. Thanks a lot for the detailed guidance offline.
:pray:

**Review by ngimel — APPROVED** [review]

**Comment by jjsjann123** [comment]
Ah, we are breaking our own tests :sob: 
I'll fix it later

**Pull request closed without merging**